### PR TITLE
Implementing Teatime effect

### PR DESCRIPTION
**Pull request by SonikkuA-DatH**
Note: requires the Bugbite/Stuffed Cheeks PR [here](https://github.com/rh-hideout/pokeemerald-expansion/pull/1581)

Teatimeeo

Solves [#1059](https://github.com/rh-hideout/pokeemerald-expansion/issues/1059)

This PR implements Teatime's effect. Has an edge case for Electrify -> Lightningrod to enable the move to cycle through all battlers with Lightning Rod/Motor Drive/Volt Absorb, and invulnerable mons

TODO [✓]: Confirm if there's ways to block Teatime outside Max Guard. Embargo and Protect don't
-There isn't other than no berry failing, Crafty Shield doesn't protect either
TODO [✓]: Confirm if ground types are affected if Teatime is Electrified
-Teatime is just considered status, works

Discord: Dat.H A#697

### CODE REVIEW

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]
I thought I had submitted this review, but I found it as pending. I'm really sorry :(

File: data/battle_scripts_1.s [hunk 1]
```diff
@@ -403,6 +403,73 @@ gBattleScriptsForMoveEffects::
...
+jumpifmotoraffected BS_TARGET, BattleScript_Teatimemotor
+	orword gHitMarker, HITMARKER_NO_ANIMATIONS
+	setbyte sBERRY_OVERRIDE, TRUE   @ override the requirements for eating berries
+	consumeberry BS_TARGET, TRUE  @ consume the berry, then restore the item from changedItems
+	bicword gHitMarker, HITMARKER_NO_ANIMATIONS
+	setbyte sBERRY_OVERRIDE, FALSE
+	moveendto MOVEEND_NEXT_TARGET
+	jumpifnexttargetvalid BattleScript_TeatimeLoop
+	moveendcase MOVEEND_CLEAR_BITS
+    goto BattleScript_MoveEnd
```
[the remark this answers is not in the document]
Comment (SonikkuA-DatH) [reply data/battle_scripts_1.s]: Should be fixed

File: data/battle_scripts_1.s [hunk 2]
```diff
@@ -403,6 +403,73 @@ gBattleScriptsForMoveEffects::
...
+    playstatchangeanimation BS_TARGET, BIT_SPATK, STAT_CHANGE_BY_TWO
+	setstatchanger STAT_SPATK, 1, FALSE
+	statbuffchange STAT_BUFF_ALLOW_PTR, BattleScript_TeatimeBuffer
+	jumpifbyte CMP_EQUAL, cMULTISTRING_CHOOSER, 0x2, BattleScript_TeatimeBuffer
+	printfromtable gStatUpStringIds
+	waitmessage 0x40
+	moveendto MOVEEND_NEXT_TARGET
+	jumpifnexttargetvalid BattleScript_TeatimeLoop
+	moveendcase MOVEEND_CLEAR_BITS
+    goto BattleScript_MoveEnd
```
[the remark this answers is not in the document]
Comment (SonikkuA-DatH) [reply data/battle_scripts_1.s]: All set

File: data/battle_scripts_1.s [hunk 3]
```diff
@@ -403,6 +403,73 @@ gBattleScriptsForMoveEffects::
...
+BattleScript_Teatimemotor:
+	copybyte gBattlerAbility, gBattlerTarget
+	call BattleScript_AbilityPopUp
+    playstatchangeanimation BS_TARGET, BIT_SPEED, STAT_CHANGE_BY_TWO
```
[the remark this answers is not in the document]
Comment (SonikkuA-DatH) [reply data/battle_scripts_1.s]: It's good now

File: data/battle_scripts_1.s [hunk 4]
```diff
@@ -403,6 +403,73 @@ gBattleScriptsForMoveEffects::
...
+	setstatchanger STAT_SPEED, 1, FALSE
+	statbuffchange STAT_BUFF_ALLOW_PTR, BattleScript_TeatimeBuffer
+	jumpifbyte CMP_EQUAL, cMULTISTRING_CHOOSER, 0x2, BattleScript_TeatimeBuffer
+	printfromtable gStatUpStringIds
+	waitmessage 0x40
+	moveendto MOVEEND_NEXT_TARGET
+	jumpifnexttargetvalid BattleScript_TeatimeLoop
+	moveendcase MOVEEND_CLEAR_BITS
+    goto BattleScript_MoveEnd
```
[the remark this answers is not in the document]
Comment (SonikkuA-DatH) [reply data/battle_scripts_1.s]: Okie dokie, it is done now

File: data/battle_scripts_1.s [hunk 5]
```diff
@@ -403,6 +403,73 @@ gBattleScriptsForMoveEffects::
...
+	ppreduce
+	jumpifteanoberry BattleScript_ButItFailed
+@ at least one battler is affected 
+	attackanimation
+	waitanimation
+BattleScript_TeatimeLoop:
+jumpifteainvulnerable BS_TARGET, BattleScript_Teatimevul
+jumpifrodaffected BS_TARGET, BattleScript_Teatimerod
+jumpifabsorbaffected BS_TARGET, BattleScript_Teatimesorb
+jumpifmotoraffected BS_TARGET, BattleScript_Teatimemotor
```
[the remark this answers is not in the document]
Comment (SonikkuA-DatH) [reply data/battle_scripts_1.s]: Easy

File: data/battle_scripts_1.s [hunk 6]
```diff
@@ -403,6 +403,73 @@ gBattleScriptsForMoveEffects::
...
 	.4byte BattleScript_EffectClangorousSoul          @ EFFECT_CLANGOROUS_SOUL
 	.4byte BattleScript_EffectHit                     @ EFFECT_BOLT_BEAK
+	.4byte BattleScript_EffectTeatime                 @ EFFECT_TEATIME
+
+BattleScript_EffectTeatime::
+	attackcanceler
+	attackstring
+	ppreduce
+	jumpifteanoberry BattleScript_ButItFailed
+@ at least one battler is affected
```
[the remark this answers is not in the document]
Comment (SonikkuA-DatH) [reply data/battle_scripts_1.s:413]: Anything else?

File: data/battle_scripts_1.s [hunk 7]
```diff
@@ -403,6 +403,73 @@ gBattleScriptsForMoveEffects::
...
+	setbyte sBERRY_OVERRIDE, FALSE
+	moveendto MOVEEND_NEXT_TARGET
+	jumpifnexttargetvalid BattleScript_TeatimeLoop
+	moveendcase MOVEEND_CLEAR_BITS
+    goto BattleScript_MoveEnd
+BattleScript_Teatimevul:
+	moveendto MOVEEND_NEXT_TARGET
+	jumpifnexttargetvalid BattleScript_TeatimeLoop
+	moveendcase MOVEEND_CLEAR_BITS
+    goto BattleScript_MoveEnd
```
[the remark this answers is not in the document]
Comment (SonikkuA-DatH) [reply data/battle_scripts_1.s]: Aaand, done!

File: src/battle_script_commands.c [hunk 8]
```diff
@@ -9285,6 +9351,49 @@ static void Cmd_various(void)
...
+			gBattlescriptCurrInstr = T1_READ_PTR(gBattlescriptCurrInstr + 3);
+		else
+			gBattlescriptCurrInstr += 7;
+		return;
+	case VARIOUS_JUMP_IF_ABSORB:
+		if (IsAbilityAbsorbAffected())
+			gBattlescriptCurrInstr = T1_READ_PTR(gBattlescriptCurrInstr + 3);
+		else
+			gBattlescriptCurrInstr += 7;
+		return;
```
[the remark this answers is not in the document]
Comment (SonikkuA-DatH) [reply src/battle_script_commands.c]: No tabs here!

File: data/battle_scripts_1.s [hunk 9]
```diff
@@ -403,6 +403,73 @@ gBattleScriptsForMoveEffects::
...
+	waitmessage 0x40
+	moveendto MOVEEND_NEXT_TARGET
+	jumpifnexttargetvalid BattleScript_TeatimeLoop
+	moveendcase MOVEEND_CLEAR_BITS
+    goto BattleScript_MoveEnd
+BattleScript_TeatimeBuffer:
+	moveendto MOVEEND_NEXT_TARGET
+	jumpifnexttargetvalid BattleScript_TeatimeLoop
+	moveendcase MOVEEND_CLEAR_BITS
+    goto BattleScript_MoveEnd
```
[the remark this answers is not in the document]
Comment (SonikkuA-DatH) [reply data/battle_scripts_1.s]: This good?

**Comment by AsparagusEduardo** [comment]
It's not building.

File: data/battle_scripts_1.s [hunk 10]
```diff
@@ -416,6 +416,73 @@ gBattleScriptsForMoveEffects::
 	.4byte BattleScript_EffectDoubleShock             @ EFFECT_DOUBLE_SHOCK
 	.4byte BattleScript_EffectSpecialAttackUpHit      @ EFFECT_SPECIAL_ATTACK_UP_HIT
 	.4byte BattleScript_EffectVictoryDance            @ EFFECT_VICTORY_DANCE
+  .4byte BattleScript_EffectTeatime                 @ EFFECT_TEATIME
```
Comment (LOuroboros) [inline data/battle_scripts_1.s:419]:
```suggestion
	.4byte BattleScript_EffectTeatime                     @ EFFECT_TEATIME
```

File: src/battle_script_commands.c [hunk 11]
```diff
@@ -8129,6 +8129,73 @@ static bool32 IsRototillerAffected(u32 battlerId)
...
+		moveType = gBattleStruct->dynamicMoveType & 0x3F;
+	else
+		moveType = gBattleMoves[gCurrentMove].type;
+
+	if (moveType == TYPE_ELECTRIC && GetBattlerAbility(gBattlerTarget) == ABILITY_LIGHTNING_ROD) {
+		return TRUE;
+	}
+	else {
+		return FALSE;
+	}
```
Comment (LOuroboros) [inline src/battle_script_commands.c:8149]:
```suggestion
	if (moveType == TYPE_ELECTRIC && GetBattlerAbility(gBattlerTarget) == ABILITY_LIGHTNING_ROD)
		return TRUE;
	else
		return FALSE;
```

File: src/battle_script_commands.c [hunk 12]
```diff
@@ -8129,6 +8129,73 @@ static bool32 IsRototillerAffected(u32 battlerId)
...
+	else
+		moveType = gBattleMoves[gCurrentMove].type;
+
+	if (moveType == TYPE_ELECTRIC && GetBattlerAbility(gBattlerTarget) == ABILITY_MOTOR_DRIVE) {
+		return TRUE;
+	}
+	else {
+		return FALSE;
+	}
+}
```
Comment (LOuroboros) [inline src/battle_script_commands.c:8169]:
```suggestion
static bool32 IsAbilityMotorAffected(void)
{
    u32 moveType;

    if (gBattleStruct->dynamicMoveType == 0)
        moveType = gBattleMoves[gCurrentMove].type;
    else if (!(gBattleStruct->dynamicMoveType & 0x40))
        moveType = gBattleStruct->dynamicMoveType & 0x3F;
    else
        moveType = gBattleMoves[gCurrentMove].type;

    if (moveType == TYPE_ELECTRIC && GetBattlerAbility(gBattlerTarget) == ABILITY_MOTOR_DRIVE)
        return TRUE;
    else
        return FALSE;
}
```

File: src/battle_script_commands.c [hunk 13]
```diff
@@ -10068,6 +10135,49 @@ static void Cmd_various(void)
...
+                }
+                }
+                if (count == 0) {
+                    gBattlescriptCurrInstr = T1_READ_PTR(gBattlescriptCurrInstr + 3);   // Teatime fails
+                }
+                else {
+                    gBattlescriptCurrInstr += 7;
+                }
+        }
+        return;
```
Comment (LOuroboros) [inline src/battle_script_commands.c:10156]:
```suggestion
    case VARIOUS_TEATIME_TARGETS:
    {
        u32 count = 0;

        for (i = 0; i < gBattlersCount; i++)
        {
            if (IsTeatimeAffected(i))
                count++;
        }
        if (count == 0)
            gBattlescriptCurrInstr = T1_READ_PTR(gBattlescriptCurrInstr + 3);   // Teatime fails
        else
            gBattlescriptCurrInstr += 7;
    }
    return;
```

**Review by LOuroboros — CHANGES_REQUESTED** [review]

**Review by LOuroboros — CHANGES_REQUESTED** [review]

File: src/battle_util.c [hunk 14]
```diff
@@ -418,7 +418,8 @@ void HandleAction_UseMove(void)
             battlerAbility = GetBattlerAbility(gActiveBattler);
 
             RecordAbilityBattle(gActiveBattler, gBattleMons[gActiveBattler].ability);
-            if (battlerAbility == ABILITY_LIGHTNING_ROD)
+
+            if (battlerAbility == ABILITY_LIGHTNING_ROD && gCurrentMove != MOVE_TEATIME)
                 gSpecialStatuses[gActiveBattler].lightningRodRedirected = TRUE;
```
Comment (AsparagusEduardo) [inline src/battle_util.c:423]:
```suggestion
            if (battlerAbility == ABILITY_LIGHTNING_ROD && gCurrentMove != MOVE_TEATIME)
                gSpecialStatuses[gActiveBattler].lightningRodRedirected = TRUE;
```

**Review by AsparagusEduardo — CHANGES_REQUESTED** [review]

**Comment by AsparagusEduardo** [comment]
Did you end up confirming the TODOs?

**Review by AsparagusEduardo — APPROVED** [review]

**Pull request merged**